### PR TITLE
FALCON-1105 REST API support for Server side extensions artifact repository management

### PR DESCRIPTION
**Pull request by sowmyaramesh**
REST API support for Server side extensions artifact repository management.
Fixed the warnings in the file.

### CODE REVIEW

**Comment by bvellanki** [comment]
:+1:

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java [hunk 1]
```diff
@@ -74,6 +79,33 @@
...
+    private static final String EXTENSION_PROPERTY_JSON_SUFFIX = "-properties.json";
+    private static final String SHORT_DESCRIPTION = "shortDescription";
+
+    private static final String TRUSTED_EXTENSION = "Trusted extension";
+    private static final String CUSTOM_EXTENSION = "Custom extension";
+
+    /**
+     * Enum for extension element.
+     */
+    private static enum ExtensionElement {
```
Comment (yzheng-hortonworks) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: The elements in this enum class are used as string constant. Looks better to just define them as constant strings.

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java [hunk 2]
```diff
@@ -74,6 +79,33 @@
     public static final String DESCENDING_SORT_ORDER = "desc";
 
     private Extension extension = new Extension();
+    private static final ExtensionStore STORE = ExtensionStore.get();
+    private static final String RESULTS = "extensions";
```
Comment (yzheng-hortonworks) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: Looks better to use "EXTENSION_RESULT" than "RESULT".

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java [hunk 3]
```diff
@@ -322,6 +354,112 @@ public APIResult validate(
...
+    @Path("enumerate")
+    @Produces({MediaType.APPLICATION_JSON})
+    public Response getExtensions() {
+        JSONArray results;
+
+        try {
+            List<String> extensions = STORE.getExtensions();
+            results = buildEnumerateResult(extensions);
+        } catch (StoreAccessException e) {
+            LOG.error("Enumerate extensions failed", e);
```
Comment (yzheng-hortonworks) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: Looks better to have a log message specific to this type of exception if you want to distinguish this exception from others, e.g. Failed when accessing extension store.

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java [hunk 4]
```diff
@@ -332,20 +470,20 @@ public APIResult validate(
         // add tags on extension name and job
         for (Entity entity : entities) {
             String tags = entity.getTags();
-            if (StringUtils.isEmpty(tags)) {
-                setEntityTags(entity, TAG_PREFIX_EXTENSION_NAME + extensionName + TAG_SEPARATOR
-                        + TAG_PREFIX_EXTENSION_JOB + properties.getProperty(ExtensionProperties.JOB_NAME.getName()));
-            } else {
-                if (tags.indexOf(TAG_PREFIX_EXTENSION_NAME) != -1) {
+            if (!StringUtils.isEmpty(tags)) {
```
Comment (yzheng-hortonworks) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: Can use StringUtils.isNotEmpty.

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java [hunk 5]
```diff
@@ -322,6 +354,112 @@ public APIResult validate(
...
+        try {
+            JSONObject jsonObject = new JSONObject(content);
+            description = (String) jsonObject.get(SHORT_DESCRIPTION);
+        } catch (JSONException e) {
+            throw new FalconException(e);
+        }
+        return description;
+    }
+
+    private static String getResource(final String extensionName, final String resourceName) throws FalconException {
```
Comment (yzheng-hortonworks) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: Looks better to keep this (getResource) and getShortDescription in ExtensionStore :)

**Comment by yzheng-hortonworks** [comment]
Could you add tests or create a jira to add tests? Thank you.

**Comment by sowmyaramesh** [comment]
Created https://issues.apache.org/jira/browse/FALCON-1921 for IT tests

File: prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java (hunk 5 above)
Comment (sowmyaramesh) [inline prism/src/main/java/org/apache/falcon/resource/extensions/ExtensionManager.java]: I will move getResource to store but would keep getShortDescription as I don't want to move JSON part to store. Extension store is mainly for operations related to accessing artifacts. Thanks!

**Comment by yzheng-hortonworks** [comment]
+1

**Pull request closed without merging**